sensor/common/deduper/deduper.go

```diff
@@ -1,43 +1,128 @@
 package deduper
 
 import (
+	"fmt"
 	"reflect"
+	"strings"
 
+	"github.com/pkg/errors"
 	"github.com/stackrox/rox/generated/internalapi/central"
 	"github.com/stackrox/rox/pkg/alert"
 	"github.com/stackrox/rox/pkg/logging"
+	eventPkg "github.com/stackrox/rox/pkg/sensor/event"
 	"github.com/stackrox/rox/pkg/sensor/hash"
 	"github.com/stackrox/rox/sensor/common/managedcentral"
 	"github.com/stackrox/rox/sensor/common/messagestream"
 )
 
 var (
 	log = logging.LoggerForModule()
+
+	deduperTypes = []any{
+		&central.SensorEvent_NetworkPolicy{},
+		&central.SensorEvent_Deployment{},
+		&central.SensorEvent_Pod{},
+		&central.SensorEvent_Namespace{},
+		&central.SensorEvent_Secret{},
+		&central.SensorEvent_Node{},
+		&central.SensorEvent_NodeInventory{},
+		&central.SensorEvent_ServiceAccount{},
+		&central.SensorEvent_Role{},
+		&central.SensorEvent_Binding{},
+		&central.SensorEvent_ProcessIndicator{},
+		&central.SensorEvent_ProviderMetadata{},
+		&central.SensorEvent_OrchestratorMetadata{},
+		&central.SensorEvent_ImageIntegration{},
+		&central.SensorEvent_ComplianceOperatorResult{},
+		&central.SensorEvent_ComplianceOperatorProfile{},
+		&central.SensorEvent_ComplianceOperatorRule{},
+		&central.SensorEvent_ComplianceOperatorScanSettingBinding{},
+		&central.SensorEvent_ComplianceOperatorScan{},
+	}
 )
 
+// Key is an interface to the key by which the messages are deduped.
+type Key interface {
+	GetID() string
+	GetType() reflect.Type
+}
+
 // key is the key by which messages are deduped.
 type key struct {
 	id           string
 	resourceType reflect.Type
 }
 
+func (k key) GetID() string {
+	return k.id
+}
+
+func (k key) GetType() reflect.Type {
+	return k.resourceType
+}
+
+func keyFrom(v string) (Key, error) {
+	parts := strings.Split(v, ":")
+	if len(parts) != 2 {
+		return key{}, fmt.Errorf("invalid Key format: %s", v)
+	}
+	t, err := mapType(parts[0])
+	if err != nil {
+		return key{}, errors.Wrap(err, "map type")
+	}
+	return key{
+		id:           parts[1],
+		resourceType: t,
+	}, nil
+}
+
+func mapType(typeStr string) (reflect.Type, error) {
+	for _, t := range deduperTypes {
+		if typeStr == eventPkg.GetEventTypeWithoutPrefix(t) {
+			return reflect.TypeOf(t), nil
+		}
+	}
+	return nil, fmt.Errorf("invalid type: %s", typeStr)
+}
+
 // deduper takes care of deduping sensor events.
 type deduper struct {
 	stream   messagestream.SensorMessageStream
-	lastSent map[key]uint64
+	lastSent map[Key]uint64
```

Comment:
Interfaces can be `nil` so there is a risk for a panic when doing map operations with a `Key` type. For that reason, a `struct` with primitive values makes more sense to me as a map key.

Reply:
Makes sense. I've changed it in #8078